feat(python): Add syntactic sugar for `col("foo")` -> `col.foo`

stinodego opened the pull request:
Closes #10866, closes #7398

#### Changes
* Implement `col` as a factory object rather than a function. It supports `col.foo` as an alias for `col("foo")`.

alexander-beedie commented:
> @alexander-beedie Is this how you would implement this functionality? I don't really see a way to satisfy mypy here without making `col` subclass `Expr`, but I think that will have some bad side effects 🤔

Not quite; can simplify - no need for the metaclass.
I can update this with a tweak that gets rid of all the `mypy` errors  ;)

<img width="541" alt="Screenshot 2023-09-02 at 18 10 04" src="https://github.com/pola-rs/polars/assets/2613171/5dfd6e39-5f2e-487c-9f26-509d2a10d2fc">

alexander-beedie commented:
Done; made the concise `c` and `cs` shortcuts available too:
```
from polars import c, cs

str(c.colx)
'col("colx")'

cs.by_name("colx", "coly")
cs.by_name('colx', 'coly')
```

ritchie46 commented:
I am not really sure a about pre-determining the shortcuts for users. Especially single char c is something will collide a lot.

I think we should just document our idiomatic important names.

stinodego commented:
> modified the implementation

Thanks! I missed the idea of instantiating the object and exporting that instead. That makes it a lot cleaner indeed.

I noticed the `_Expr_` object wasn't necessary, so I removed it. And I'm also not sure about exporting `c` and `cs` directly. I'll leave it out of this PR for now.

I'm also wondering what your intention is with naming the helper class `_column_`? Default Python style `Column` seems fully appropriate here. It will be exposed to the user in the end.

I switched `__getattr__` to `__getattribute__` as we are certain there are no other things to look up on this class.

### Idiomatic usage

We should think about what we think idiomatic usage should look like here, and what we want to use in our own docs.
I'd like to adhere to a single idiomatic style ourselves. Do we just keep using `pl.col("foo")` in all our docstrings / book examples?

alexander-beedie commented:
> I think we should just document our idiomatic important names.

Yup, makes sense - `c` turns up too much to just claim it as our own :)

> I noticed the _Expr_ object wasn't necessary

Oh, nice - in my first iteration `mypy` complained hundreds of times about `col` not being callable, heh; must have accidentally fixed it n the second iteration without noticing :)

> I'm also wondering what your intention is with naming the helper class _column_?

Habit; anything I don't want the users instantiating themselves I generally name such that they can't claim they didn't know, haha. I'd still probably have "col" in an explicit `__all__` though.

> I switched __getattr__ to __getattribute__ 

😎👍

ritchie46 commented:
> We should think about what we think idiomatic usage should look like here, and what we want to use in our own docs.
I'd like to adhere to a single idiomatic style ourselves. Do we just keep using pl.col("foo") in all our docstrings / book examples? I added a comment to the linked issue to discuss this.

I want to strongly push for keeping that. It is the most complete. It allows for all possible names in a DataFrame. For instance spaces and `-` are pretty common in header names.

stinodego commented:
> I want to strongly push for keeping that.

Great, then we're in agreement!

>  I'd still probably have "col" in an explicit `__all__`

Yeah on second thought I agree - added it back!

alexander-beedie commented:
Hmm, just tested the latest updates and I think the move to use `__getattribute__` is probably a bit too aggressive; it blows-up a lot of typical object-introspection interaction from IDEs (for instance, if you hit `.` in PyCharm it will immediately throw an error stack at you:

<img width="492" alt="Screenshot 2023-09-03 at 17 05 10" src="https://github.com/pola-rs/polars/assets/2613171/6bfeead1-90e7-4436-905c-49ddfb10638f">

stinodego commented:
Yeah, considering this is not intended as a new idiomatic way of doing things, but rather for quick prototyping that's not intended for supporting every single use case, `getattr` is probably fine.

Meanwhile, I am down the rabbithole of getting things to look nice in Sphinx... 🙃

alexander-beedie commented:
World's smallest nitpick - rename `Column` as `ColumnFactory`?
It creates columns, but definitely isn't one itself.

stinodego commented:
> It creates columns, but definitely isn't one itself.

That's true - renamed to `ColumnFactory`!

ritchie46 reviewed (commented):
Nice, thanks guys!
